### PR TITLE
[ENH] Add support for AAHC clustering

### PR DESCRIPTION
**Pull request by rkobler**
Hi pycrostates team @vferat @mscheltienne ! 👋 

Thank you for working on a python repository that integrates microstate analysis with MNE python.

I recently started using the repository and felt like adding support for the atomize and agglomerate hierarchical clustering (AAHC) algorithm. The implementation is inspired by the matlab code provided in the eeglab plugin Microstates v 1.2 [1].

Please, let me know if you see merit in adding this method to the package.

Best,
Reinmar

[1] https://www.thomaskoenig.ch/Download/EEGLAB_Microstates/Microstates1.2.zip

### CODE REVIEW

**Comment by mscheltienne** [comment]
Thank you very much for adding this algo! I just went once through the code quickly, it looks very good. I will try to test it quickly :)

For the style, you can run `black` at the root of the repository with the command `black .` or `black pycrostates` and it will auto-format the files, fixing most of the style issues.

Also, could you add an entry in the changelog: https://github.com/vferat/pycrostates/tree/main/docs/source/dev/changes
You can add your name and a link to your website/github in `names.inc` and an entry for this PR in `latest.rst` 🙏

File: pycrostates/cluster/aahc.py [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+
+                    v0 = old_weight * sgn * old_cluster + \
+                        (1. - old_weight) * cluster[:, c]
+
+                    v, _ = AAHCluster._first_principal_component(
+                        data[:, members], tol, v0
+                    )
+                    cluster[:, c] = v
+                else:
+                    cluster[:, c] = np.mean(data[:, members], axis=0)
```
Comment (vferat) [inline pycrostates/cluster/aahc.py:278]: It seems that the mean is made on the wrong axis (`axis=0` instead of `axis=1`)

`cluster[:, c] = np.mean(data[:, members], axis=1)`

Otherwise, an error is raised we trying to fit `AAHCluster `with `ignore_polarity=False`

Would you have time to also add test(s) to cover this case ?

File: pycrostates/cluster/aahc.py [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+        if self.fitted:
+            n_samples = self._fitted_data.shape[-1]
+            ch_types, ch_counts = np.unique(
+                self.get_channel_types(), return_counts=True
+            )
+            ch_repr = [
+                f"{ch_count} {ch_type.upper()}"
+                for ch_type, ch_count in zip(ch_types, ch_counts)
+            ]
+            GEV = int(self._GEV_ * 100)
```
Comment (vferat) [inline pycrostates/cluster/aahc.py:93]: I feel like the convertion to interger could lead to missunderstanding ( GEV = 50.56% but displays 50%).
Maybe something like `GEV = f'{self._GEV_ * 100:.2f}'` would be more suitable.

I see that `ModKMeans `has the same issue. Feel free to correct it for both Classes, otherwise I'll do it in another PR.

**Review by vferat — COMMENTED** [review]
Hey @rkobler,

Thanks a lot for this contribution, we really appreciate it.  As @mscheltienne  said, the code is very clean. This a a great addition to the library.
I checked your code with some data, and only find one/two small issues. Please have a look at my review for more details.

**Comment by mscheltienne** [comment]
I have one question about the argument `ignore_polarity`. The ModKMeans always ignore polarities, right?
Why is this an option with AAHC and what is a use case of `ignore_polarity=False`?

File: pycrostates/cluster/__init__.py [hunk 3]
```diff
@@ -15,5 +15,6 @@
 to segment."""
 
 from .kmeans import ModKMeans  # noqa: F401
+from .aahc import AAHCluster
```
Comment (mscheltienne) [inline pycrostates/cluster/__init__.py:18]:
```suggestion
from .aahc import AAHCluster  # noqa: F401
```

It's just to avoid warnings in IDEs ;)

File: pycrostates/cluster/aahc.py [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+    def fit(
+        self,
+        inst: Union[BaseRaw, BaseEpochs],
+        picks: Picks = "eeg",
+        tmin: Optional[Union[int, float]] = None,
+        tmax: Optional[Union[int, float]] = None,
+        reject_by_annotation: bool = True,
+        n_jobs: int = 1,
+        *,
+        verbose: Optional[str] = None,
```
Comment (mscheltienne) [inline pycrostates/cluster/aahc.py:156]: @vferat `verbose` should probably be defined and set in the base-class. Which will remove from both ModKMeans and this class the `fill_doc` decorator around the `fit` method and its docstring.

File: pycrostates/cluster/aahc.py [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+        tol: Union[int, float],
+    ) -> Tuple[NDArray[float], NDArray[int]]:
+        """Compute microstates maps."""
+        n_chan, n_frame = data.shape
+
+        cluster = data.copy()
+        cluster /= np.linalg.norm(cluster, axis=0, keepdims=True)
+
+        if normalize_input:
+            data = cluster.copy()
```
Comment (mscheltienne) [inline pycrostates/cluster/aahc.py:236]: Could you add a test with `normalize_input=True`?

File: pycrostates/cluster/aahc.py [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+        Parameters
+        ----------
+        X : numpy.ndarray
+            Input data matrix with dimensions (channels x observations)
+        tol : float
+            Tolerance for convergence.
+        v0 : numpy.ndarray or None (default)
+            Initial estimate of the principal vector with dimensions
+            (channels, )
+            If None, a random initialization is used
```
Comment (mscheltienne) [inline pycrostates/cluster/aahc.py:307]:
```suggestion
        v0 : numpy.ndarray or None (default)
            Initial estimate of the principal vector with dimensions
            (channels,). If None, a random initialization is used.
```

File: pycrostates/cluster/aahc.py [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+
+                    v, _ = AAHCluster._first_principal_component(
+                        data[:, members], tol, v0
+                    )
+                    cluster[:, c] = v
+                else:
+                    cluster[:, c] = np.mean(data[:, members], axis=0)
+                    cluster[:, c] /= np.linalg.norm(
+                        cluster[:, c], axis=0, keepdims=True
+                    )
```
Comment (mscheltienne) [inline pycrostates/cluster/aahc.py:281]: I was looking at how `tol` was used, and how it defines convergence. Note: I am not familiar with AAHC.

It looks like:
- `tol` is not used if `ignore_polarity=False`. That's weird, right?
- There is no information/fix possible in case the convergence was not reached. I think a warning should be emitted if convergence was not reached and `AAHCluster._first_principal_component` has a `max_iter` argument which should be forwarded to the public API, to enable the user to increase it if need be.

File: pycrostates/cluster/aahc.py [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+
+        Returns
+        -------
+
+        tuple(v, eig)
+
+        v : numpy.ndarray
+            estimated principal component vector
+        eig : float
+            eigenvalue of the covariance matrix of X (channels x channels)
```
Comment (mscheltienne) [inline pycrostates/cluster/aahc.py:320]:
```suggestion
        v : numpy.ndarray
            Estimated principal component vector.
        eig : float
            Eigenvalue of the covariance matrix of X (channels x channels)
```

I'm going to guess `pydocstyle` would complain otherwise.

File: pycrostates/cluster/aahc.py [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+        X : numpy.ndarray
+            Input data matrix with dimensions (channels x observations)
+        tol : float
+            Tolerance for convergence.
+        v0 : numpy.ndarray or None (default)
+            Initial estimate of the principal vector with dimensions
+            (channels, )
+            If None, a random initialization is used
+        max_iter : int
+            Maximum number of iterations to estiamte the first principal
```
Comment (mscheltienne) [inline pycrostates/cluster/aahc.py:309]:
```suggestion
            Maximum number of iterations to estimate the first principal
```

File: pycrostates/cluster/aahc.py [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+        v : numpy.ndarray
+            estimated principal component vector
+        eig : float
+            eigenvalue of the covariance matrix of X (channels x channels)
+
+        See :footcite:t:`Roweis1997` for additional information.
+        """
+
+        if v0 is None:  # use a random choice
+            v = np.random.rand(X.shape[0])
```
Comment (mscheltienne) [inline pycrostates/cluster/aahc.py:326]: It looks like `v0` is always provided, thus I would remove this condition and edit the arguments type-hint and docstring.

File: pycrostates/utils/_checks.py [hunk 11]
```diff
@@ -67,6 +67,7 @@ def __instancecheck__(cls, other):
     "path-like": (str, Path, os.PathLike),
     "int": (_IntLike(),),
     "callable": (_Callable(),),
+    "bool": (bool,),
```
Comment (mscheltienne) [inline pycrostates/utils/_checks.py:70]: Why did you need this?

`_check_type(True, (bool,))` should work directly, no need for a string wrapper.

File: pycrostates/cluster/aahc.py (hunk 1 above)
Comment (rkobler) [reply pycrostates/cluster/aahc.py:278]: Thank you for the positive feedback. I have not tested the case `ignore_polarity=False`. I will fix the error and add a test case.

File: pycrostates/cluster/aahc.py (hunk 5 above)
Comment (rkobler) [reply pycrostates/cluster/aahc.py:236]: Yes, I will add a test case.

File: pycrostates/utils/_checks.py (hunk 11 above)
Comment (rkobler) [reply pycrostates/utils/_checks.py:70]: You are right. I reverted this change and updated the corresponding calls in aahc.py

File: pycrostates/cluster/aahc.py (hunk 1 above)
Comment (rkobler) [reply pycrostates/cluster/aahc.py]: Resolved in https://github.com/vferat/pycrostates/pull/92/commits/a3f7afc062b2f43cf61a9bc2b43f7af29a470524

File: pycrostates/cluster/aahc.py (hunk 2 above)
Comment (rkobler) [reply pycrostates/cluster/aahc.py]: Resolved in https://github.com/vferat/pycrostates/pull/92/commits/1161e8eeba3a0c522d5e5577454d45423393527a

File: pycrostates/cluster/aahc.py (hunk 5 above)
Comment (rkobler) [reply pycrostates/cluster/aahc.py:227]: Resolved in https://github.com/vferat/pycrostates/pull/92/commits/a3f7afc062b2f43cf61a9bc2b43f7af29a470524

File: pycrostates/cluster/aahc.py (hunk 7 above)
Comment (rkobler) [reply pycrostates/cluster/aahc.py]: Thank you for the comment. Yes, `tol` was only used in the case `ignore_polarity=True`. In this case, AAHC requires to compute the first principal component (PC) of the cluster members. There are two options. First, power iterations (as implemented), which is fast but requires some heuristic to determine convergence. Second, conventional PCA which is slow because it requires the covariance matrix and then computes all eigenvectors/values.

In commit  https://github.com/vferat/pycrostates/pull/92/commits/a3f7afc062b2f43cf61a9bc2b43f7af29a470524, I removed the `tol` parameter from the public API. In case the fast power iteration method does not converge within a reasonable iteration (e.g., number of features = number of channels), the fallback method is used (covariance matrix + largest eigenvalue/vector).

File: pycrostates/cluster/aahc.py (hunk 10 above)
Comment (rkobler) [reply pycrostates/cluster/aahc.py]: Resolved in commit https://github.com/vferat/pycrostates/pull/92/commits/a3f7afc062b2f43cf61a9bc2b43f7af29a470524

File: pycrostates/cluster/aahc.py (hunk 5 above)
Comment (rkobler) [reply pycrostates/cluster/aahc.py:227]: Resolved in commit https://github.com/vferat/pycrostates/pull/92/commits/29fa0e53ffd2a1aefc633a32e47bec430afcb275

**Comment by rkobler** [comment]
@mscheltienne 

> I have one question about the argument `ignore_polarity`. The ModKMeans always ignore polarities, right?
Why is this an option with AAHC and what is a use case of `ignore_polarity=False`?

This is a valid question. I think the best motivation for including the case `ignore_polarity=False` are scenarios where one would like to apply clustering to event-related potential (ERP) datasets. In this case, the polarity is informative.

**Comment by vferat** [comment]
> @mscheltienne 
> 
> > I have one question about the argument `ignore_polarity`. The ModKMeans always ignore polarities, right?
> Why is this an option with AAHC and what is a use case of `ignore_polarity=False`?
> 
> This is a valid question. I think the best motivation for including the case `ignore_polarity=False` are scenarios where one would like to apply clustering to event-related potential (ERP) datasets. In this case, the polarity is informative.

@rkobler is right. For ERP analysis we take into account polarities whereas we ignore it for spontaneous activities ( raw / epochs).

I did not intend to include ERP microstates analysis in pycrostates at first. But it may be interesting. For now, I think we should keep the `ignore_polarity` hidden from the public API  and always use the default value `True`.

 `Metrics`, `.predict()`  are only implemented to work with ignored polarities. I may lead to wrong results if people try to use them with a cluster that do not ignore polarities.

In the meantime, I'll have a look at how much work is needed to implement changes to take into account polarities in the library.

**Comment by mscheltienne** [comment]
Ok, then yes +1 to include `ignore_polarity` in ModKMeans as well in a different PR. And indeed, if for now `predict` does not work when polarities are not ignored, then let's hard-code `self._ignore_polarity = True` in `AAHC.__init__()` with a TODO item to move it back as an argument when we add support for `ignore_polartity=False` in the base-class, in ModKMeans, and in the metrics.

File: pycrostates/cluster/aahc.py (hunk 7 above)
Comment (mscheltienne) [reply pycrostates/cluster/aahc.py]: Thank you, looks great, I love this fallback ;)

File: pycrostates/cluster/aahc.py [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+        normalize_input: bool,
+    ) -> Tuple[NDArray[float], NDArray[int]]:
+        """Compute microstates maps."""
+        n_chan, n_frame = data.shape
+
+        cluster = data.copy()
+        cluster /= np.linalg.norm(cluster, axis=0, keepdims=True)
+
+        if normalize_input:
+            data = cluster.copy()
```
Comment (mscheltienne) [inline pycrostates/cluster/aahc.py:227]:
```suggestion
        cluster = data.copy()

        if normalize_input:
            cluster /= np.linalg.norm(cluster, axis=0, keepdims=True)
```

Shouldn't it be this? Or am I missing something?

File: pycrostates/cluster/aahc.py [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+                        Cxx = data[:, members] @ data[:, members].T
+                        _, V = np.linalg.eigh(Cxx)
+                        v = V[:, -1]
+
+                    cluster[:, c] = v
+                else:
+                    cluster[:, c] = np.mean(data[:, members], axis=1)
+                    cluster[:, c] /= np.linalg.norm(
+                        cluster[:, c], axis=0, keepdims=True
+                    )
```
Comment (mscheltienne) [inline pycrostates/cluster/aahc.py:281]: One more question about this part, I'm a bit surprised by the difference between the steps when `ignore_polarity` is set to True and the steps when it is set to False. 
Could you expand a bit on those, either here, or even maybe in a couple of comments in the code?

Mostly, I'm trying to build confidence that it shouldn't be:

```
if ignore_polarity:
    # do some extra-steps

cluster[:, c] = np.mean(data[:, members], axis=1)
cluster[:, c] /= np.linalg.norm(cluster[:, c], axis=0, keepdims=True)
```

Also, do you know if there is a scikit-learn implementation of this algorithm (which would be the case `ignore_polarity=False`)? I found `sklearn.cluster.AgglomerativeClustering` but I'm not sure it's the same algo.

Comment (mscheltienne) [reply pycrostates/cluster/aahc.py:281]: Nevermind, @vferat convinced me.

File: pycrostates/cluster/aahc.py (hunk 12 above)
Comment (rkobler) [reply pycrostates/cluster/aahc.py:227]: The version that I submitted is consistent with the reference implementation [1] (eeg_computeAAHC.m lines 15 to 19). That is, the clusters *and* data are normalized.

[1] https://www.thomaskoenig.ch/Download/EEGLAB_Microstates/Microstates1.2.zip

File: pycrostates/cluster/aahc.py (hunk 4 above)
Comment (rkobler) [reply pycrostates/cluster/aahc.py:155]: I agree with @mscheltienne . Should I implement this change in this PR?

File: pycrostates/cluster/aahc.py (hunk 12 above)
Comment (mscheltienne) [reply pycrostates/cluster/aahc.py:237]: I completely misread those lines, I thought this copy wasn't achieving anything, but it does replace the argument `data` with itself normalized.

File: pycrostates/cluster/aahc.py (hunk 4 above)
Comment (mscheltienne) [reply pycrostates/cluster/aahc.py:155]: If you feel like it sure, else I'll do it after.

File: pycrostates/cluster/aahc.py [hunk 14]
```diff
@@ -0,0 +1,397 @@
...
+
+    # --------------------------------------------------------------------
+
+    @property
+    def ignore_polarity(self) -> bool:
+        """If true, polarity is ignored when computing distances.
+
+        :type: `bool`
+        """
+        return self._ignore_polarity
```
Comment (mscheltienne) [inline pycrostates/cluster/aahc.py:359]: This property should be removed for now, until we add proper support.

File: pycrostates/cluster/aahc.py [hunk 15]
```diff
@@ -0,0 +1,397 @@
...
+
+                    v, _, converged = AAHCluster._first_principal_component(
+                        data[:, members], v0, max_iter=n_chan
+                    )
+                    if not converged:
+                        # fall back to covariance estimation
+                        # and eigenvalue computation
+                        Cxx = data[:, members] @ data[:, members].T
+                        _, V = np.linalg.eigh(Cxx)
+                        v = V[:, -1]
```
Comment (mscheltienne) [inline pycrostates/cluster/aahc.py:284]: Could you add a logger line at the debug level, to be able to check how often we have to fall back?
I would even add the `else` statement with another debug line to check how often we don't have to.

File: pycrostates/cluster/tests/test_aahc.py [hunk 16]
```diff
@@ -0,0 +1,1358 @@
...
+    # Trust MNE .get_data() to correctly select data
+    data = epochs.get_data(picks=picks, tmin=tmin, tmax=tmax)
+    # check channels
+    assert fitted_data.shape[0] == data.shape[1]
+    # check samples
+    assert fitted_data.shape[1] == int(data.shape[0] * data.shape[2])
+
+
+def test_aahClusterMeans():
+    """Test K-Means default functionalities."""
```
Comment (mscheltienne) [inline pycrostates/cluster/tests/test_aahc.py:226]:
```suggestion
    """Test AAHC default functionalities."""
```

File: pycrostates/cluster/tests/test_aahc.py [hunk 17]
```diff
@@ -0,0 +1,1358 @@
...
+    assert aah_cluster._fitted_data is not None
+    assert aah_cluster._info is not None
+    assert aah_cluster.GEV_ is not None
+    assert aah_cluster._labels_ is not None
+
+
+def _check_unfitted(aah_cluster):
+    """
+    Checks that the aah_cluster is not fitted.
+    """
```
Comment (mscheltienne) [inline pycrostates/cluster/tests/test_aahc.py:188]:
```suggestion
    """Check that the aah_cluster is not fitted."""
```

File: pycrostates/cluster/tests/test_aahc.py [hunk 18]
```diff
@@ -0,0 +1,1358 @@
...
+    assert (dists.min(axis=0) < tol).all()
+    # ensure that all cluster centers were identified
+    assert len(set(dists.argmin(axis=0))) == sim_n_ms
+
+
+# pylint: disable=protected-access
+def _check_fitted(aah_cluster):
+    """
+    Checks that the aah_cluster is fitted.
+    """
```
Comment (mscheltienne) [inline pycrostates/cluster/tests/test_aahc.py:174]:
```suggestion
    """Check that the aah_cluster is fitted."""
```

File: pycrostates/cluster/tests/test_aahc.py [hunk 19]
```diff
@@ -0,0 +1,1358 @@
...
+    #         n_clusters=n_clusters,
+    #         ignore_polarity=None
+    #     )
+
+    # normalize_input
+    with pytest.raises(
+        TypeError, match="'normalize_input' must be an instance of bool"
+    ):
+        aahCluster_ = AAHCluster(n_clusters=n_clusters, normalize_input="asdf")
+        aahCluster_ = AAHCluster(n_clusters=n_clusters, normalize_input=None)
```
Comment (mscheltienne) [inline pycrostates/cluster/tests/test_aahc.py:637]: Unless I am mistaken, the `pytest.raises` context manager will only catch the first matching exception and then skip the rest of the nested code. Take the `foo` function:

```
def foo(x):
    if x == 1:
        raise RuntimeError("test")
    else:
        return x
```

If you run:

```
import pytest

with pytest.raises(RuntimeError, match="test"):
    x = (foo(5))
    foo(1)
```

The variable `x` is defined at the end. But if you run:

```
import pytest

with pytest.raises(RuntimeError, match="test"):
    foo(1)
    x = (foo(5))
```

The variable `x` is not defined at the end. That means you can not test multiple calls that should raise the same error message in the same context manager, or only the first one will actually be tested.

**Review by mscheltienne — APPROVED** [review]
Only minor comments remain, LGTM. 
I especially like the tests on simulated data you added!

File: pycrostates/cluster/tests/test_aahc.py (hunk 19 above)
Comment (rkobler) [reply pycrostates/cluster/tests/test_aahc.py]: Thanks for pointing me on this issue. I updated the code accordingly.

**Review by mscheltienne — APPROVED** [review]
LGTM, thanks a lot for this amazing addition, and don't worry about the codacy workflow ;)

**Comment by mscheltienne** [comment]
Good catch for `n_jobs`, @vferat I pushed an additional commit to fix the docstring, since the order of the arguments would have been wrong.

**Comment by vferat** [comment]
Thx @mscheltienne for the fix !

LGTM ! @rkobler thanks a lot for the contribution!  Everything went super smooth thanks to your code quality and fast responses 🥇 

@all-contributors please add @rkobler  for code, doc, ideas and test !
